metals/src/main/scala/scala/meta/internal/metals/TargetData.scala

```diff
@@ -126,24 +132,69 @@ final class TargetData {
   def info(id: BuildTargetIdentifier): Option[BuildTarget] =
     buildTargetInfo.get(id)
 
+  /**
+   * Get jars for a specific build target.
+   *
+   * We first try to use buildTargetDependencyModules
+   * request since it should be low cost for build tools
+   * like Bazel.
+   *
+   * We fall back to reading from classpath only if the
+   * classpath is read eagerly.
+   *
+   * @param id id of the queried target
+   * @return depenendency jar list if available
+   */
   def targetJarClasspath(
       id: BuildTargetIdentifier
   ): Option[List[AbsolutePath]] = {
-    val scalacData = scalaTarget(id).map(_.scalac.jarClasspath)
-    val javacData = javaTarget(id).map(_.javac.jarClasspath)
-    scalacData
-      .flatMap(s => javacData.map(j => (s ::: j).distinct).orElse(scalacData))
-      .orElse(javacData)
+    buildTargetDependencyModules.get(id) match {
+      case None =>
+        buildTargetClasspath.get(id)
+        scalaTarget(id).flatMap { target =>
+          target.lazyJarClasspath
+        }
+      case Some(value) =>
+        Some(value.flatMap(_.getArtifacts().asScala).collect {
+          case artifact: MavenDependencyModuleArtifact
+              if artifact.getClassifier() == null =>
+            artifact.getUri().toAbsolutePath
+        })
+    }
   }
 
   def targetClasspath(
       id: BuildTargetIdentifier
-  ): Option[List[String]] = {
-    val scalacData = scalaTarget(id).map(_.scalac.classpath)
-    val javacData = javaTarget(id).map(_.javac.classpath)
-    scalacData
-      .flatMap(s => javacData.map(j => (s ::: j).distinct).orElse(scalacData))
-      .orElse(javacData)
+  )(implicit ec: ExecutionContext): Option[Future[List[String]]] = {
+    targetToConnection.get(id).zip(scalaTarget(id)).map {
```

Comment:
Do we completely not care about `javaTargets` here? This is also used for resolving classpath for java pc.

Reply:
Added a common trait for both JavaTarget and ScalaTarget and now it will just get whichever exists